q2_fmt/_peds.py

```diff
@@ -466,11 +479,205 @@ def _create_masking(time_metadata, donor_df, recip_df, reference_column):
+                                 used_references,
+                                 reference_column)
+    donor_mask = _create_sim_masking(mismatched_df, donor_df, reference_column)
+    mismatched_pairs_n = len(duplicated_table.index)
+    recip_mask = donor_mask & duplicated_table
+    # Numerator for PEDS Calc. (Number of Donor features in the Recipient)
+    num_sum = np.sum(recip_mask.values, axis=1)
+    # Denominator for PEDS Calc. (Number of unique features in the Donor)
+    donor_sum = np.sum(donor_mask, axis=1)
+    # This ignores warnings that come from dividing by 0.
+    # mismatched_peds will be Nan if the denominator is 0 and thats reasonable.
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        mismatched_peds = num_sum/donor_sum
+    mismatchedpairs_df = _simulate_uniform_distro(recip_df, mismatched_peds,
+                                                  iterations)
+    per_sub_stats = _per_subject_stats(actual_temp, mismatchedpairs_df,
+                                       iterations, mismatched_pairs_n)
+    global_stats = _global_stats(per_sub_stats['p-value'])
+    return per_sub_stats, global_stats
+
+
+# peds_simulation helper functions
+
+def _create_mismatched_pairs(recip_df, metadata, used_references,
+                             reference_column):
+    # creates mismatched_pairs of donors and recipients.
+    # Also creates a recipient feature table where there is a duplicate
+    # recipient sample for every pair.
+    # This makes numpy array math much simpler later.
+    mismatched_pairs = []
+    for x in itertools.product(recip_df.index,
+                               metadata[reference_column].dropna().unique()):
+        mismatched_pairs.append(x)
+    mismatched_t = list(zip(used_references.index, used_references))
+    # Removes matched donor and recipient pairs
+    filtered = [pair for pair in mismatched_pairs if pair not in mismatched_t]
+    idx, values = zip(*filtered)
+    mismatched_df = pd.DataFrame({"id": idx,
+                                  reference_column: values}).set_index("id")
+    duplicated_table = \
+        mismatched_df.merge(recip_df, left_index=True, right_index=True,
+                            how='left').drop(mismatched_df.columns, axis=1)
+    return duplicated_table, mismatched_df
+
+
+def _create_sim_masking(mismatched_df, donor_df, reference_column):
+    # Creates a Donor Numpy array that duplicates donor samples to match up
+    # with duplicated_table. This will mask recipient feature that aren't
+    # in the donor.
+    donors = mismatched_df[reference_column]
+    donor_index_masking = []
+    for donor in donors:
+        donor_index_masking.append(donor_df.index.get_loc(donor))
+    donor_df_np = donor_df.to_numpy()
+    donor_mask = donor_df_np[donor_index_masking]
+    donor_mask = donor_mask.astype(int)
+    return donor_mask
+
+
+def _simulate_uniform_distro(recip_df, mismatched_peds, iterations):
```

Comment:
Rather than create all of these upfront and store them in memory, create each one when you need it (likely in your `_per_subject_stats` function. Since you only need each of these one time (and one at a time) that will reduce unnecessary memory usage.